### PR TITLE
[Debugger Plugin]: Add Source Code View

### PR DESCRIPTION
**Pull request by caisq**
Main feature:
* Viewing the Python source code files involved in the creation of nodes
  of the graph. (Only files outside the TensorFlow library are
  included.)
* Lines of the code files are annotated with collapsible views of the
  nodes whose creation stacks involve the lines.
* Clicking the node names in the Source Code View will focus on the
  corresponding node in the Node List, the Graph View and the Tensor
  Value Overview.
* Clicking the "forward" icon button next to the node name in the
  Source Code View will cause the debugger to continue to the node and
  display any tensor values in the Tensor Value Overview.
* Clicking the "stack" icon button right next to the node name will
  open a sticky dialog that shows the full stack trace of the node's
  creation. The stack trace contains clickable links that enables
  the user to navigate to various frames of the stack, potentially
  located in different source files.

Front end tested manually. Backend PyUnit tests added.

### CODE REVIEW

**Comment by caisq** [comment]
Example screenshot:
tdp2-source-3

**Comment by jart** [comment]
Source code view in debugger? Now that's really cool. Some high level feedback:

1. We might want a flag in main.py to turn this off. That flag is a promise to users that TensorBoard will never show source code, in this plugin, or any other. Once disabled, it should be OK for debugger to just show filename, line number, and function name.

2. What happens when I click the unusually bright blue underlined text? Does it show the whole source file? What does it look like? Would it be sufficient to just show the ±5 lines of the statement currently being evaluated? I ask because I want to make sure whatever we dip our toes into, we do it really really well.

**Comment by caisq** [comment]
@jart Thanks for the comments. To address your questions:

1. I submitted a change on the tensorflow side to allow the debugged TF runtime to disable the sending
    of source code and graph tracebacks. I think it's a more reliable solution than disabling source code /
    traceback display on the tensorboard side. It should help address cases in which source code
    confidentiality and/or the network communication overhead associated with the code and traceback
    data is of concern.

2. Clicking the blue underlined links in the line-by-line node list in the source code will do the
     following:
     * Focus on the corresponding node in the Node List (top left quadrant)
     * Focus on the corresponding node in the Graph Visualizer (top right quadrant)
     * Highlight the corresponding tensor in the Tensor Value Overview (bottom right quadrant)
    Clicking the blue underlined links in the full stack sticky dialog will scroll the source code to the 
    corresponding line, and switch to a different source code file first if necessary.

Currently the full content of a source code file is displayed. I think it's simpler than displaying a small
number of lines at a time. The ability to inspect any part of a file is nice. The network overhead should
be acceptable, given that most user files are reasonably sized (<1000 lines, or about 8 kB). Note that we
currently omit files that belong to the TF python library itself (i.e., see the grayed out items in the
full stack sticky dialog).

**Comment by jart** [comment]
That sounds OK to me.

File: tensorboard/plugins/debugger/tf_debugger_dashboard/tf-source-code-view.html [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+        }
+        // Then, create all the clickable links.
+        for (const stackFrame of response['op_traceback'][nodeName]) {
+          const frameElement = document.createElement('li');
+          const fullFilePath = stackFrame[0];
+          const lineno = Number(stackFrame[1]);
+          frameElement.textContent = fullFilePath + ': ' + String(lineno);
+          if (_.contains(this._fullFilePaths, fullFilePath)) {
+            frameElement.classList.add('stack-frame-clickable');
+            // TODO(cais): Make class CSS work.
```
Comment (chihuahua) [inline tensorboard/plugins/debugger/tf_debugger_dashboard/tf-source-code-view.html]: See other comment.

Comment (caisq) [reply tensorboard/plugins/debugger/tf_debugger_dashboard/tf-source-code-view.html]: See other reply. Unfortunately adding Polymer.dom() still doesn't work for me. I hope we can look into it together sometime this wee.

**Comment by caisq** [comment]
Thanks for the review, @chihuahua!

**Comment by caisq** [comment]
Comments addressed. Rebasing on master done.

**Pull request merged**